onnx/defs/math/defs.cc

```diff
@@ -734,6 +729,76 @@ static const char* MatMul_ver6_doc = R"DOC(
 Matrix product that behaves like numpy.matmul: https://docs.scipy.org/doc/numpy-1.13.0/reference/generated/numpy.matmul.html
 )DOC";
 
+void matmulShapeInference(InferenceContext& ctx) {
+  if (!hasNInputShapes(ctx, 2)) {
+    return;
+  }
+
+  const auto shape0 = ctx.getInputType(0)->tensor_type().shape();
+  const auto shape1 = ctx.getInputType(1)->tensor_type().shape();
+
+  if (shape0.dim_size() == 0 || shape1.dim_size() == 0) {
+    fail_shape_inference("Input tensors of wrong rank (0).");
+  }
+
+  TensorShapeProto shapeL, shapeR;
+
+  // First promote each shape to at least rank-2. This logic is
+  // specific to matmul, not generic broadcasting.
+  {
+    if (shape0.dim_size() == 1) {
+      shapeL.add_dim()->set_dim_value(1);
+      *shapeL.add_dim() = shape0.dim(0);
+    } else {
+      *shapeL.mutable_dim() = shape0.dim();
+    }
+    if (shape1.dim_size() == 1) {
+      *shapeR.add_dim() = shape1.dim(0);
+      shapeR.add_dim()->set_dim_value(1);
+    } else {
+      *shapeR.mutable_dim() = shape1.dim();
+    }
+  }
+
+  // Check for compatible matrix multiply dimensions
+  {
+    auto dimL = shapeL.dim(shapeL.dim_size() - 1);
```

Comment:
Nit: commonDimL?

Reply:
This is not that common, I think :), so will keep the name. Again, I didn't touch this function, but make it to be used by both matmul and matmulinteger.